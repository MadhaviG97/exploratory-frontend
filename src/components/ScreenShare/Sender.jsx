import React from 'react';
import '../../assets/css/screenshare.css';
import openSocket from 'socket.io-client';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import InputLabel from '@material-ui/core/InputLabel';
import Input from '@material-ui/core/Input';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import { withStyles } from "@material-ui/core/styles";
import axios from 'axios';
import Avatar from '@material-ui/core/Avatar';
import Box from '@material-ui/core/Box';
const socket = openSocket('http://localhost:8000');
var caller;
//var group="GeeFour"
var callername="caller"
//var clients = io.sockets.clients('room');
//var room=group.concat(caller)
/*{
        url: 'turn:numb.viagenie.ca',
        credential: 'muazkh',
        username: 'webrtc@live.com'
      }
      {
      "urls": [
      "turn:13.250.13.83:3478?transport=udp"
      ],
      "username": "YzYNCouZM1mhqhmseWk6",
      "credential": "YzYNCouZM1mhqhmseWk6"
      }*/ 
const RTC_CONFIGURATION = {
    iceServers: [
      { urls: 'stun:stun.l.google.com:19302' },
      {
        urls: 'turn:webrtcweb.com:80',
        username: 'muazkh',
        credential: 'muazkh'
      },
      {
        urls: 'turn:webrtcweb.com:443',
        username: 'muazkh',
        credential: 'muazkh'
      }
    ]
  };
const styles = theme => ({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  formControl: {
      margin: theme.spacing(1),
      minWidth: 290,
      maxWidth: 300,
  },
});

  function sendOffer(offer,id) {
    socket.emit('offer', {room: id,offer:offer});
    
  }
  function sendCandidate(candidate,id,user) {
    let sender=''
    console.log(user)
    if (user.userData){
      sender=user.userData.first_name.concat(' ').concat(user.userData.last_name)
    }
    socket.emit('candidate', {room: id,candidate:candidate});
    socket.emit('sender', {sender:sender});
  }
  function getStream() {
    return new Promise((resolve, reject) => {
      navigator.mediaDevices.getDisplayMedia({ audio: false, video: true }).then((mediaStream) => {
        resolve(mediaStream);
      })
    });
  }
  function makePeerConnection(stream,id,user) {
    if (caller) { caller.close() }
    caller = new RTCPeerConnection(RTC_CONFIGURATION);
    caller.addStream(stream);
    caller.onicecandidate = (event) => {
      if (event.candidate != null) {
        sendCandidate(event.candidate,id,user);
      }
    }
    makeOffer(id)
  }
  function makeOffer(id) {
    caller.createOffer().then((offer) => {
      return caller.setLocalDescription(offer);
    }).then(() => {
      sendOffer(caller.localDescription,id);
    });
  }
const researchers= [ {name: 'mocked name 1',_id:'10001'},{name: 'mocked name 2',_id:'1235'} ]
class Sender extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        open:false,
        researcher:{},
        collaborators:[]
    };
    this.ITEM_HEIGHT = 48;
    this.ITEM_PADDING_TOP = 8;
    this.group=props.group
    this.MenuProps = {
      PaperProps: {
        style: {
          maxHeight: this.ITEM_HEIGHT * 4.5 + this.ITEM_PADDING_TOP,
          width: 300,
        },
      },
    };  
  }
  
  handleClickOpen = () => {
    this.setState({
      open:true});
  };

  handleClose = () => {
    this.setState({
      open:false});
  };
  handleChange = (event) => {
    this.setState({
      researcher:event.target.value});
    console.log(event.target.value)
  };
    
    
    componentDidMount() {
        
        const variable = { 
          group:this.group
        
        }
        console.log(variable)
        /*
        const token = localStorage.token;
        let config = {
            headers: {
            'Authorization': `Bearer ${token}`
            }
          }
        
        */
        axios.post('/project/get-collaborators', variable)
            .then(response => {
                if (response.data) {
                  console.log(response.data)
                    this.setState({collaborators:response.data})
                    console.log(response.data[0].profile_picture)
                }
            })
            
        socket.on('answer', (answer) => {
            caller.setRemoteDescription(answer);
        });
        socket.on('candidate', (candidate) => {
            caller.addIceCandidate(candidate);
        });

        
    }

    
    handleSubmit=(user)=>{
        //console.log(this.state)
        this.setState({
            open:false});
        var room=this.state.researcher.researcher_id.toString().concat('caller')
        socket.emit('join', room);
        getStream().then((stream) => {
            document.getElementById('screen').srcObject = stream;
            makePeerConnection(stream,this.state.researcher.researcher_id.toString(),user);
        });
        
    }
    
    
    render() {
      let user=this.props.user
      const { classes } = this.props;
        return (
            <div>
                <Dialog open={this.state.open} onClose={this.handleClose} aria-labelledby="form-dialog-title">
                  <DialogContent>
                  <form className={classes.container}>
                    <FormControl className={classes.formControl}>
                      <InputLabel id="demo-dialog-select-label">Select a Receiver</InputLabel>
                      <Select
                        labelId="demo-dialog-select-label"
                        id="demo-dialog-select"
                        value={this.state.researcher}
                        onChange={this.handleChange}
                        input={<Input />}
                        MenuProps={this.MenuProps}
                      >
                      {this.state.collaborators.filter(researcher => researcher.researcher_id !== user.userData._id).map((researcher) => (
                        <MenuItem key={researcher.researcher_id} value={researcher} >
                          <Avatar alt="user" src={researcher.pofile_picture} />
                          <Box p={1}/>
                          {researcher.researcher_email}
                        </MenuItem>
                      ))}
                      </Select>
                  </FormControl>
                  </form>
                  </DialogContent>
                  <DialogActions>
                    <Button onClick={this.handleClose} color="primary">
                      Cancel
                    </Button>
                    <Button onClick={() => {
                            this.handleSubmit(user);
                        }} 
                        color="primary">
                      Connect
                    </Button>
                  </DialogActions>
                </Dialog>
                <video controls playsInline autoPlay id="screen"></video>
                
                <div style={{ textAlign: 'center', margin: '2rem', }}>
                    <Button
                        size="large"
                        htmlType="submit"
                        style={{  background: '#014f82',//can change the sign-in button color from here
                        color: '#FFFFFF',
                        height: 40,
                        boxShadow: ['none']}}
                        variant="contained"
                        component="label"
                        onClick={this.handleClickOpen}
                        //onSubmit={onSubmit}
                        >
                            Share the Screen
                    </Button>    
                </div>
            </div>
        )
    }

}

export default withStyles(styles)(Sender);
