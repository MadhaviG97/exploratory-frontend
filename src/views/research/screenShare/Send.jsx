import React from "react";
import classNames from "classnames";
import { useEffect } from "react";
import axios from "axios";
import Box from "@material-ui/core/Box";
import { useStyles } from "../../../assets/css/projectFolderGrid";
import Sender from "../../../components/ScreenShare/Sender";
import { useSelector } from "react-redux";
import NavComponent from "../../../components/AppNavigation/NavigationComponent";
import Divider from "@material-ui/core/Divider";
import NotFound from "../../../components/NotFound/NotFound";
import Loader from "../../../components/Loader";

export default function Receive(props) {
  const user = useSelector((state) => state.user);
  console.log(user);
  const classes = useStyles();
  let user_id = 0;
  if (user.userData) {
    user_id = user.userData._id;
  }
  const [collabs, setCollabs] = React.useState([]);
  useEffect(() => {
    const variable = {
      group: props.match.params.projectId,
    };
    axios.post("/project/get-collaborators", variable).then((response) => {
      if (response.data) {
        setCollabs(response.data);
      }
    });
  }, []);
  if (user.userData) {
    if (collabs.some((e) => e.researcher_id == user_id)) {
      return (
        <div className={classNames(classes.main2)}>
          <div style={{ width: "70%", margin: "1rem auto" }}>
            <NavComponent projectId={props.match.params.projectId} />

            <Divider variant="fullWidth" />
            <Box p={0.5} />
            {/*
                            <Box boxShadow={3} style={{  background: '#FFFFFF'}} >
                                <Box p={0.5}>
                                    <div >
                                        <h1 align='center' className={classes.topic}>Share Your Screen with Project members</h1>
                                    </div>
                                </Box>
                            </Box>
                       
                        */}
          </div>
          <div>
            <Sender group={props.match.params.projectId} user={user} />
          </div>
          <Box p={5} />
        </div>
      );
    } else {
      return <NotFound />;
    }
  } else {
    return <Loader />;
  }
}
