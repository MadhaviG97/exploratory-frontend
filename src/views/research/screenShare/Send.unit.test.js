import React from 'react';
import ReactDOM from 'react-dom';
import  Send  from './Send';
import {shallow,mount} from 'enzyme'
import Sender from "../../../components/ScreenShare/Sender"
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import {Provider} from 'react-redux'
import renderer from 'react-test-renderer';

const realUseState = React.useState
const stubInitialState = [{researcher:{researcher_id:'10002'}}]

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
  useLocation: () => ({
    push: jest.fn(),
  }),
}));

const mockStore = configureMockStore([thunk]);
const match = { params: { projectId: '10012' } }
const store = mockStore({
    user: { userData: {isAuth:true,first_name:'yogya',id:'10002'} }
  });
const collabs=[{researcher:{researcher_id:'10002'}}]

describe('Share Screens', () => {

    it('renders Sender Component', () => {
      jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => realUseState(stubInitialState))
      .mockImplementationOnce(() => realUseState(true))
      const wrapper = mount(
      <Provider store={store}>
        <Send  match={match}/>
      </Provider>);
      expect(wrapper.find(Sender).exists()).toBeTruthy();
    });

    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
      <Provider store={store}>
          <Send match={match}/>
      </Provider>, div);
  });
  
    it('renders correctly', () => {
        const tree = renderer
          .create(
          <Provider store={store}>
            <Send  match={match}/>
          </Provider>)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });