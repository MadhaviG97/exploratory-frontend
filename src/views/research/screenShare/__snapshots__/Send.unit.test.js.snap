// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share Screens renders correctly 1`] = `
<div>
  <main
    className="MuiGrid-root makeStyles-root-138 MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-5 MuiGrid-grid-md-5"
    />
    <paper
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 MuiGrid-grid-md-2"
      elevation={6}
      square={true}
    >
      <div
        className="MuiGrid-root MuiGrid-container"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <circle
            color="#014F82"
            size={150}
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <circle
            color="#014F82"
            size={150}
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <circle
            color="#014F82"
            size={150}
          />
        </div>
      </div>
    </paper>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-5 MuiGrid-grid-md-5"
    />
  </main>
  <div
    className="MuiBox-root MuiBox-root-244"
  />
</div>
`;
